src/Clients/ClientBase.cs

```diff
@@ -0,0 +1,64 @@
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using StreamChat.Exceptions;
+using StreamChat.Models;
+using StreamChat.Rest;
+using HttpMethod = StreamChat.Rest.HttpMethod;
+
+namespace StreamChat.Clients
+{
+    public abstract class ClientBase
+    {
+        private readonly IRestClient _client;
+
+        internal ClientBase(IRestClient client)
+        {
+            _client = client;
+        }
+
+        protected async Task<T> ExecuteRequestAsync<T>(
+            string relativeUri,
+            HttpMethod method,
+            HttpStatusCode expectedStatusCode,
+            object body = null,
+            IEnumerable<KeyValuePair<string, string>> queryParams = null,
+            MultipartFormDataContent multipartBody = null) where T : ApiResponse
+        {
+            var req = _client.BuildRestRequest(relativeUri, method, body, queryParams, multipartBody);
+
+            var resp = await _client.ExecuteAsync(req);
+
+            if (resp.StatusCode == expectedStatusCode)
+            {
+                T deserialized;
+
+                try
+                {
+                    deserialized = JsonConvert.DeserializeObject<T>(resp.Content);
+                }
+                catch (Exception ex)
+                {
+                    throw new StreamDeserializationException("Failure during response deserialization.", ex);
```

Comment:
This doesn't give much extra information to the user who catches this exception, would be best to also pass typeof(T) and resp.Content.

Scenario I have in mind:
```
            try
            {
                await _channelClient.AddModeratorsAsync(_channel.Type, _channel.Id, new[] { _user2.Id });
            }
            catch (Exception e)
            {
                //breakpoint here
            }
```
you use the Api client and it fails -> you catch exception and inspect it with debugger.

And what you'd like to know from exception is what string was tried to be deserialized to which exact type so you can figure out why it failed. Right now you don't get this information from exception only, you need to follow the stack trace to figure out which method's serialization failed and from that check type and string.

Reply:
good point